### ISSUE TITLE
QApplication.quit breaks following tests

### ISSUE THREAD

**Issue by rdevries-thoughtworks** [pytest-dev/pytest-qt#323]
I have a test that uses `waitCallback`, which succeeds when run in isolation, but fails if a `QApplication.quit` is run from my code somewhere before. In that case, it fails immediately (not after the timeout), with a `TimeoutError`.

(Sorry I don't have time (now) to give more details, might that be necessary)

**Comment by The-Compiler** [comment]
I don't think this is pytest-qt's fault - if you run `QApplication.quit` but then later run tests which require a QApplication (and anything relying on the Qt mainloop, including singlas, do), all bets are off.

**Comment by rdevries-thoughtworks** [comment]
Hi @The-Compiler , I don't want to blame anyone. ;-) If you'd create a separate QApplication for every test, though, that would fix this issue, right? I'm not sure if there's any (performance) issues with that, but otherwise having isolated tests is super valuable, and shared state between tests generally a huge pain to deal with.

**Comment by The-Compiler** [comment]
As far as I'm aware, a QApplication object can't be reliably destroyed/recreated. You'd probably need to run every test in a separate process with something like [pytest-boxed](https://github.com/pytest-dev/pytest-forked).

**Comment by rdevries-thoughtworks** [comment]
Ouch! I see. Would it make sense (and be possible) to just crash the tests or fail with a clear message if the QApplication has been quit? Then at least debugging becomes easier.

**Comment by The-Compiler** [comment]
Perhaps it'd be possible to connect to the `aboutToQuit` signal and somehow exit/fail in that case - I'm not entirely sure, though. Certainly worth to try.

**Comment by BoniLindsley** [comment]
There is an undocumented [shutdown][] method that can be used, before creating another `QCoreApplication`. You can try it out with this:

[shutdown]: https://doc.qt.io/qtforpython/PySide2/QtCore/QCoreApplication.html#PySide2.QtCore.PySide2.QtCore.QCoreApplication.shutdown

```py
#!/usr/bin/env python3
from PySide2.QtCore import QCoreApplication as Application
Application().shutdown()
Application().shutdown()
```

An [example] of this in unit tests. Making this work in more cases is... [fiddly][] to say the least. I have to stress again that it is [undocumented] though. (Also be warned that, if you try to make this work, there seems to be some `QCoreApplication` [inheritance] issue in Python 3.9, causing my `QTestApplication` to crash in Debian.)

[example]: https://gist.github.com/BoniLindsley/eb0f5c13c46f32f9629853a0119e250b
[fiddly]: https://github.com/BoniLindsley/phile/blob/a26c336e49894b813d28e177fc71e32deaf56feb/test_phile/pyside2_test_tools.py
[added]: https://bugreports.qt.io/browse/PYSIDE-1190?focusedCommentId=494586#comment-494586
[undocumented]: https://gist.github.com/BoniLindsley/96da34b7bf2d0faffa5a9a59baf0e66f
[inheritance]: https://gist.github.com/BoniLindsley/96da34b7bf2d0faffa5a9a59baf0e66f
